network-config-analyzer/CanonicalHyperCubeSet.py

```diff
@@ -2,367 +2,732 @@
+                remaining_self_layer -= common_elem
+                if self._is_last_dimension():
+                    # do not add common_elem to self.layers here because result is empty
+                    continue
+                # sub-elements subtraction
+                new_sub_elem = (self.layers[self_layer].copy()).sub_aux(other.layers[other_layer])
+                if bool(new_sub_elem):
+                    # add remaining new_sub_elem if not empty, under common
+                    res_layers[common_elem] = new_sub_elem
+            if remaining_self_layer:
+                res_layers[remaining_self_layer] = self.layers[self_layer]
+        self.layers = res_layers
+        self._apply_layer_elements_union()
+        return self
 
+    def _prepare_common_active_dimensions(self, other):
+        """
+        change self and other so that they have common active dimensions.
+        Changes to 'other' should be on a new copy, not directly on it.
+        :type other: CanonicalHyperCubeSet
+        :rtype CanonicalHyperCubeSet
+        :return: result for 'other' (which may be copied and changed or not)
+        """
+        if self.active_dimensions == other.active_dimensions:
+            return other
+        required_active_dimensions = set(self.active_dimensions + other.active_dimensions)
+        self._set_active_dimensions(required_active_dimensions)
+        if set(other.active_dimensions) == required_active_dimensions:
+            return other
+        # should not change active dimensions for 'other' during the computation of operation between self and other
+        other_copy = other.copy()
+        other_copy._set_active_dimensions(required_active_dimensions)
+        return other_copy
+
+    def is_all(self):
+        """
+        :return: bool indicating if self is the entire defined space
+        """
+        return not self.active_dimensions
+
+    def set_all(self):
+        """
+        update self to consist of a cube of the entire defined space
+        """
+        self.active_dimensions = []
+        self.layers = dict()
+
+    # TODO: use _prepare_common_active_dimensions ? (extend it?)
     def contained_in(self, other):
-        # type: (CanonicalHyperCubeSet) -> bool
-        if self.dimensions == 1:
-            return self.baseIntervalSet.contained_in(other.baseIntervalSet)
+        """
+        check containment between CanonicalHyperCubeSet objects: is self contained in other
+        :type other: CanonicalHyperCubeSet
+        :rtype bool
+        """
+        required_active_dimensions = set(self.active_dimensions + other.active_dimensions)
+        # containment check should not change active dimensions of self/other
+        self_copy = self.copy()
+        other_copy = other.copy()
+        self_copy._set_active_dimensions(required_active_dimensions)
+        other_copy._set_active_dimensions(required_active_dimensions)
+        return self_copy._contained_in_aux(other_copy)
```

Comment:
Isn't there a way to do it without copying `self` and `other`? Sounds too expensive for me.
If we copy them both, then we can simply `return bool(self - other)`

Reply:
changed the implementation